### ISSUE TITLE
Custom shrinkers like jsverify

### ISSUE THREAD

**Issue by msteen** [dubzzz/fast-check#1503]
## 💬 Question and Help
Despite loving the idea that you don't have to bother writing your own shrinkers most of the time in `fast-check`, for more complex cases it falls short. I was thinking of using a JSON diff as the counter examples produced contained too much noise to see what might have gone wrong. And rather than going that route or updating my old `jsverify` code, I thought it better to just ask how to do this in `fast-check` instead. My approach as mentioned in #1316, works good enough for most cases, but now that I have solved most issues, it seems to only produce counter examples that are not shrinked properly. I am pretty sure this is due to my use of `.chain`, but without it I cannot express how it should go about shrinking. So unless there have been some updates addressing the shrinking of recursive data structures that would allow me to do without `.chain`, I figure I will have to write my own shrinker.

In my previous issue #1316, I was told to check [ArbitraryWithShrink.ts](https://github.com/dubzzz/fast-check/blob/main/src/check/arbitrary/definition/ArbitraryWithShrink.ts) if I wanted to go about writing my own shrinker. The problem I have with this API is that for complex values, you want subvalues to use existing shrinkers, but I can't. Fundamental building blocks like `.array` don't have a `.shrinkableFor`. What then if I have an object value that I want to write a `.shrinkableFor` for, containing an array in one of its values? I don't want to reproduce the shrinking logic for arrays just because my object contains one. Ideally I would only have to write the custom shrinking logic and leverage builtins shrinkers for the rest, like you do in `jsverify`. Am I going about it the wrong way, or are there any examples of this type of shrinking?

As an example, this is what I did in my `jsverify` code to shrink `Tag` objects and what I want to reproduce in `fast-check`:
```typescript
function shrinkTag(ensureIsAttribute: boolean) {
  return shr.bless((tag: Tag) => {
    let shrs = lazyseq.nil
    if (tag.contents.length > 0) {
      shrs = shrs.append(() => {
        let tags = tag.contents.filter(content => typeof content !== "string") as Tag[]
        if (ensureIsAttribute) {
          tags = tags.map(tag => ({ ...tag, isAttribute: true }))
        }
        return tags
      })
    }
    if (tag.attributes.length > 0) {
      shrs = shrs.append(tag.attributes)
    }
    if (tag.contents.length > 0) {
      if (tag.isLiteral) {
        shrs = shrs.append(shrText(tag.contents[0] as string).map(text => ({ ...tag, contents: [text] })))
      } else {
        shrs = shrs.append(shrContents(tag.contents).map(contents => ({ ...tag, contents })))
      }
    }
    if (tag.attributes.length > 0) {
      shrs = shrs.append(
        shr
          .array(shrAttribute)(tag.attributes)
          .map(attributes => ({ ...tag, attributes })),
      )
    }
    shrs = shrs.append(shrTagName(tag.name).map(name => ({ ...tag, name })))
    if (tag.isLiteral) {
      shrs = shrs.append([{ ...tag, isLiteral: false }])
    }
    if (tag.isAttribute && !ensureIsAttribute) {
      shrs = shrs.append([{ ...tag, isAttribute: false }])
    }
    if (tag.isQuoted) {
      shrs = shrs.append([{ ...tag, isQuoted: false }])
    }
    return shrs
  })
}
```

**Comment by dubzzz** [comment]
>  So unless there have been some updates addressing the shrinking of recursive data structures that would allow me to do without `.chain`

It should come in the next minors. I drafted it locally but did not have time to send it yet. It should come next month I think.

>  Fundamental building blocks like `.array` don't have a `.shrinkableFor`

Unfortuantely yes. This one might be fixed in next major as I'm planning a huge rework on how shrinkers work internally. Contrary to jsverify `fc.array(fc.oneof(fc.nat(), fc.string())` can be shrunk. Unfortunately in order to archieve that I canntot have `.shrinkableFor` on arrays.

See https://github.com/dubzzz/fast-check/blob/main/documentation/HowItWorks.md#arbitraries

> `Tag`

What is the exact type for `Tag`? How would you produce it if recursive structures were correctly wired for shrinking?

_I just want to make sure that my future implementation might solve your issue._

**Comment by msteen** [comment]
> It should come in the next minors. I drafted it locally but did not have time to send it yet. It should come next month I think.

Awesome! I have grown to love property based testing and would not trust some of my code without them, given the edge cases I have seen and otherwise missed. However, most my input tend to consist of recursive data structures, so having them work properly is really a must for me.

> Unfortuantely yes. This one might be fixed in next major as I'm planning a huge rework on how shrinkers work internally. Contrary to jsverify `fc.array(fc.oneof(fc.nat(), fc.string())` can be shrunk. Unfortunately in order to archieve that I canntot have `.shrinkableFor` on arrays.

I get that `.shrinkableFor` won't work for `fc.oneof`, as this information is lost in the general case. I actually don't care too much about shrinking outside (i.e. not produced by combinators) values, but I did not know how to go about custom shrinking otherwise, given the current API.

Still, I don't see why you could not have `.shrinkableFor` for all, even if some, like `fc.oneof`, won't be able to do much. Even if not perfect, being able to shrink what can be shrunk would be an useful thing to have. And the API could be made so that it is easy to make `fc.oneof` work with shrinking, by adding a predicate or just in general by making it easier to write a custom shrinker that chooses how to shrink based on the generated value. I get that this is still imperfect, as again, the information might have been lost, so simply based on the generated value you might not be able to decide, but in many other cases it might be a simple `typeof` / `instanceof`.

> What is the exact type for `Tag`? How would you produce it if recursive structures were correctly wired for shrinking?

Ideally I would just define a `fc.record` with recursive fields that will properly shrink themselves too without any custom code needed. Here is my full code with `fc.boolean` and `fc.chain` tricks removed, and without the assertion (it just checks whether `print(generated) === print(parse(print(generated)))`):

```typescript
import * as fc from "fast-check"

type Tag = {
  isQuoted: boolean
  isAttribute: boolean
  name: string
  attributes: Tag[]
  isLiteral: boolean
  contents: Content[]
}

type Content = string | Tag

function joinTexts(contents: Content[]): Content[] {
  return contents.reduceRight((contents, content) => {
    if (typeof content === "string" && typeof contents[0] === "string") {
      contents[0] = content + contents[0]
    } else {
      contents.unshift(content)
    }
    return contents
  }, [] as Content[])
}

function AlphabetArbitrary(alphabet: string) {
  return fc.integer(0, alphabet.length - 1).map(i => alphabet[i])
}

function oneOutOf(n: number) {
  return fc.integer(1, n).map(i => i === n)
}

const visibleAsciiArb = AlphabetArbitrary(
  "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789!\"#$%&'()*+,-./:;<=>?@[\\]^_`{|}~",
)

const alphaArb = AlphabetArbitrary("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ")
const alphanumericArb = AlphabetArbitrary("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789")

const textArb = fc.stringOf(
  fc.frequency(
    { arbitrary: fc.constant("\\"), weight: 1 },
    { arbitrary: fc.constant("{"), weight: 3 },
    { arbitrary: fc.constant("}"), weight: 3 },
    { arbitrary: fc.constant(":"), weight: 2 },
    { arbitrary: fc.constant("\n"), weight: 1 },
    { arbitrary: visibleAsciiArb, weight: 12 },
  ),
)

const nameArb = fc
  .tuple(fc.stringOf(alphaArb, { minLength: 1 }), fc.array(fc.stringOf(alphanumericArb, { minLength: 1 })))
  .map(([s1, ss]) => [s1].concat(ss).join(" "))

const tagArb: fc.Memo<Tag> = fc.memo(n =>
  fc
    .record({
      isQuoted: oneOutOf(10),
      isAttribute: oneOutOf(3),
      name: nameArb,
      attributes: fc.array(tagArb(n - 1).map(tag => ({ ...tag, isAttribute: true }))),
      isLiteral: oneOutOf(4),
      contents: contentsArb(n - 1),
    })
    .map(tag => ({
      ...tag,
      isLiteral: tag.isLiteral && tag.contents.length === 1 && typeof tag.contents[0] === "string",
    })),
)

const contentsArb: fc.Memo<Content[]> = fc.memo(n =>
  n > 1
    ? fc
        .array(fc.oneof(textArb, tagArb(n - 1)))
        .map(contents => contents.flat())
        .map(joinTexts)
    : fc.array(textArb, { maxLength: 1 }),
)
```

A simplified example of this was given in #1316.

> _I just want to make sure that my future implementation might solve your issue._

Thank you for your dedication to this awesome project!

I might be naive here, but when I was thinking how I would go about implementing it, I would keep track of 3 things: 1) the generators that produced the values 2) the generated values 3) the final values (i.e. after things like mapping). Maybe have an additional kind of map method or some kind of marker to indicate that those values should be considered for (2), requiring them to be lossless. Then if you have a custom shrinker of a complex value (i.e. consisting of other nested generators), you would be able to access those nested value through the structure of (1) and get their values as (2), but how should be shrunk might depend on (3).

**Comment by msteen** [comment]
I don't know what you have planned to shrink recursive datatypes defined through memo, but if I still had to supply this manually, I would love it to be something like this:

```typescript
const tagArb: fc.Memo<Tag> = fc.memo(n =>
  fc
    .record({
      isQuoted: oneOutOf(10),
      isAttribute: oneOutOf(3),
      name: nameArb,
      attributes: fc.array(tagArb(n - 1).map(tag => ({ ...tag, isAttribute: true }))),
      isLiteral: oneOutOf(4),
      contents: contentsArb(n - 1),
    })
    .map(tag => ({
      ...tag,
      isLiteral: tag.isLiteral && tag.contents.length === 1 && typeof tag.contents[0] === "string",
    }))
    .withShrink(shr => {
      let shrs = Stream.nil()
      shrs = shrs.append(shr.fields.contents.filter(shr => typeof shr.value === "object"))
      shrs = shrs.append(shr.fields.attributes)
      return shrs.append(shr.shrinks())
    })
)
```

The idea behind `.withShrink` is that you can add a custom shrinking strategy while keeping the rest as normal. And every `shr` in that code keep track of both value and structure, preventing loss of information. The biggest problem I felt with shrinkers at the moment is that I felt you have two worlds, values and arbitraries, and once you get a value, you are out of the arbitraries world, losing that information. I hope you get what I mean despite being somewhat vague.

**Comment by msteen** [comment]
My biggest issue with `fast-check` at the moment is that I am not really given the means to manually write a shrinking strategy without also having to write my own shrinkers for fundamentals like `fc.array`. This makes `fast-check` not a replacement for `jsverify` for me, when really it should, as it is otherwise superior in every way. I rather write 65 lines of custom shrinkers than have 20 lines of arbitrary chain trickery that fail for edge cases (which are kinda the point of property based testers). And maybe that changes with future updates and is no longer necessary, but I don't believe it can be made perfect, so for those use cases that require custom shrinking logic, there should at least be an alternative strategy.

**Comment by dubzzz** [comment]
The philosophy behind fast-check is different from the one behind jsverify.

**The key idea for fast-check is that users should never need to do extra tricks to have something working. It should automatically detect the worst without needing to be guide to it (otherwise a simple test should be enough, no need for property based).**

This key idea has several impacts:
- by default an `fc.array(fc.nat())` generates long and small arrays, with large and small values
- map does not neep an unmap
- for shrinkers: one should not have to rewrite shrinkers or only in very rare cases

Nonetheless your point is fair: extending core shrinkers with custom logic is not simple (they have not been designed to), because they inherently deals with all the logic required for `oneof` stuff that highly complexifies the problem. Regarding to your current issue, I'll say that the best solution would be to wait for the recursive enhancement. The core revamping of arbitraries is a long task and for the moment I am not sure to be able to find a way to have something like `shrinkableFor` on `array` before a while.

May I ask you a question? - Why do you really want to rewrite your own shrinker? Even if I clearly like when shrinker reports the smallest case, for me it is just the cherry on the cake. PBT is more than just a shrinker, it is must of all a generator of combinations that can detect issues you never thought of. In a way, shrinkers are not supposed to be triggered all the time: in normal workflows, they do never appear.

**Comment by dubzzz** [comment]
Normally I should have a working version of shrinkers on recursive structures pretty soon. I'll let you know as soon as I get something usable to have your feedback on it.

Regarding `shrinkableFor` on others arbitraries, I will certainly start working on it right after. I'm planning to have it for the next major of fast-check. I'm currently working on 3.x and alpha versions should come soon (it doesn't mean that 2.x is over as I'm still planning many features on it before moving to 3.x).

**Comment by msteen** [comment]
I get the difference in philosophy and its consequences, I just hoped it to be an extra on top in terms of user facing code (I get it has big consequences in terms of implementation). Like that it would still allow everything `jsverify` and its kind do, but by being smart about it in the implementation you get this nice behavior where in most cases you can get by with just the arbitraries as they are given in the API. It is kinda like where you would have a webpage with links, but the JavaScript makes thumbnails on hover or whatever, but removes the `<a>` tag for some `window.location` trickery. It makes it nicer, but now I cannot use middle-click to open in another tab anymore, i.e. I expected the smart part of the philosophy to be an enhancement without losing access to the fundamentals. However I get what I assumed to be the fundamentals are not quite the same due to the consequences of the philosophy.

> May I ask you a question? - Why do you really want to rewrite your own shrinker?

Sure. If you mean, why rewrite `jsverify` to `fast-check` when I have something that seems to work, that is because future work will require much more property based testing and I rather not invest in a library that is unmaintained. If you mean, why aim for perfection if it is good enough, that is because it is not good enough. I get a screen full of generated gibberish where I have to use a diff tool to figure out what might have gone wrong with the counter example. Fortunately it only seemed to do this for a few bugs in my code, so I just fixed them and continued, but this will be a problem in my future properties as well, since most of them are of a similar nature.

> PBT is more than just a shrinker

The reason for my focus on shrinking is simply being confronted with overly big values of which I know 95% is noise. That it is sometimes not quite as small as it could be, is fine, but if you can no longer spot the problem/differences, the counter example is simply bad (of course even a horrible counter example is better than staying ignorant of a bug in your code).

> What you are looking for is in the pipe

That is great to hear! This issue made me held back commiting fully to `fast-check`, but seeing it addressed and taking seriously, I will write the rest of my planned tests in it as well and make it one of my standard dev libraries included in my projects.

> Normally I should have a working version of shrinkers on recursive structures pretty soon. This feature is in my to-do since 
my contribution on shrinking-challenge repo. I'll let you know as soon as I get something usable to have your feedback on it.
Thank you, if you comment on this thread I will be sure to give feedback!

> Regarding `shrinkableFor` on others arbitraries, I will certainly start working on it right after. I'm planning to have it for the next major of fast-check. Was already on my to-do as I may have found a way to deal with it. I'm currently working on 3.x and alpha versions should come soon (it doesn't mean that 2.x is over as I'm still planning many features on it before moving to 3.x).

It is nice to hear it being so actively developed. Best of luck with all that you have planned!

**Comment by dubzzz** [comment]
@msteen I POC-ed a small version that should make fast-check more efficient when dealing with recursive structures. I'm wondering if it could answer to your needs. See #1537

The package can be installed by using:
```js
yarn add https://60300bf150dc580a45751048--dazzling-goodall-4f0e38.netlify.app/fast-check.tgz
```

Additional remark: To benefit fully from all the features of the current POC it's better to use `letrec` over `memo`. `memo` will only benefit from the `withCrossShrink` flag but not from the others (as it creates new instances of the arbitraries instead of re-using them when it goes deeper).

**Comment by msteen** [comment]
To add it I had to use:
```
yarn add fast-check@https://60300bf150dc580a45751048--dazzling-goodall-4f0e38.netlify.app/fast-check.tgz
```
Due to using Yarn 2. Since I worked out all bugs in my tests and it was not easy to reintroduce some that might trigger bigger counterexamples, I tested it out by modifying my simplified test case used in #1316:
```typescript
import * as fc from "fast-check"
type Record = { contents: Content[] }
type Content = number | Record
const valueArb = fc.nat(999)
const recordArb = fc.letrec(tie => ({
  record: fc.record({
    contents: fc.frequency(
      { arbitrary: tie("contents") as fc.Arbitrary<Content[]>, weight: 1 },
      { arbitrary: fc.array(valueArb), weight: 2 },
    ),
  }),
  contents: fc.array<Content>(fc.oneof(valueArb, tie("record") as fc.Arbitrary<Record>)),
})).record
const includesBadValue = (content: Content) =>
  typeof content === "number" ? content === 666 : content.contents.some(includesBadValue)
fc.assert(fc.property(recordArb, record => !includesBadValue(record)))
```
Unfortunately it still produces counterexamples like:
```json
[{"contents":[{"contents":[{"contents":[{"contents":[666]}]}]}]}]
```

**Comment by dubzzz** [comment]
Can you try to had `{withCrossShrink: true}` as first parameter of your `fc.frequency`?

**Comment by msteen** [comment]
> Can you try to had `{withCrossShrink: true}` as first parameter of your `fc.frequency` and `fc.oneof`? It does not seem to be the case in the snippet you sent.

I tried, but I get the same result. This is what I used:
```typescript
import * as fc from "fast-check"
type Record = { contents: Content[] }
type Content = number | Record
const valueArb = fc.nat(999)
const recordArb = fc.letrec(tie => ({
  record: fc.record({
    contents: fc.frequency(
      { withCrossShrink: true },
      { arbitrary: tie("contents") as fc.Arbitrary<Content[]>, weight: 1 },
      { arbitrary: fc.array(valueArb), weight: 2 },
    ),
  }),
  contents: fc.array<Content>(fc.oneof(valueArb, tie("record") as fc.Arbitrary<Record>)),
})).record
const includesBadValue = (content: Content) =>
  typeof content === "number" ? content === 666 : content.contents.some(includesBadValue)
fc.assert(fc.property(recordArb, record => !includesBadValue(record)))
```

**Comment by dubzzz** [comment]
@msteen

I am currently working on a full rewrite of the `Arbitrary` API. In the current major (2.x), I'll include the change in a backward compatible way: any arbitrary defined by fast-check will under-the-hood use the new API while exposing the old one. Helpers functions like `convertFromNext` and `convertToNext` will make users able to use the new API and will be provided to ease migration.

Among the things that should be unlocked by this change, I'm planning to offer an API close to `bless` or `withShrink`. For the moment, I'm confirming step by step that the migration will be feasible. Once confirmed I'll start thinking about the `bless` or `withShrink`. But it will definitely be part of my POC. I'll let you know when I get something good enough to let you try and have your feedbacks on it (currently an incomplete WIP: #1631).

**Comment by dubzzz** [comment]
I have a first drafted version. For the moment, it has a poor support of existing arbitraries (only part of them have been fully migrated).

On caller-side extending an existing shrinker could be done that way:
```js
const tagArb = fc.memo((n) => (n > 0 ? fc.array(tagArb(n - 1)) : fc.nat()));
const myTagArb = fc.convertToNext(tagArb(1));
const myTagArbWithExtendedShrinker = withCustomShrinker(myTagArb, (value, originalShrink) => {
  return fc.Stream.of(...value.map((v) => new fc.NextValue(v)))
    .join(originalShrink());
});

// Or
const myTagArbWithExtendedShrinker = withCustomShrinker(myTagArb, (value) => {
  return fc.Stream.of(...value.map((v) => new fc.NextValue(v)))
    .join(myTagArb.shrink(value));
});
```

See description of the PR #1631 for the full snippet.

**Comment by dubzzz** [comment]
@msteen Here a first adapted version based on fast-check@2.15.0

```js
const fc = require('fast-check')

// Helpers

// Similar to bless but not yet in the built-ins of fast-check
function withCustomShrinker(oldSourceArb, newShrinker) {
  const sourceArb = fc.convertToNext(oldSourceArb);
  class WithCustomShrinker extends fc.NextArbitrary {
    generate(mrng) {
      return sourceArb.generate(mrng);
    }
    canShrinkWithoutContext(value) {
      return sourceArb.canShrinkWithoutContext(value);
    }
    shrink(value, context) {
      return newShrinker(value, () => sourceArb.shrink(value, context));
    }
  }
  return fc.convertFromNext(new WithCustomShrinker());
}

// Definition of the arbitraries

const valueArb = fc.nat(999)
const recordArb = fc.letrec(tie => ({
  record: withCustomShrinker(
    fc.record({
      contents: fc.frequency(
        { arbitrary: fc.array(fc.oneof(valueArb, tie("record"))), weight: 1 },
        { arbitrary: fc.array(valueArb), weight: 2 },
      ),
    }),
    (value, defaultShrink) => {
      // NOTE:
      // We will only be called if the value was produced by source
      // or if source.canShrinkWithoutContext is true for this value
      let shrs = fc.Stream.nil();
      for (let idx = 0 ; idx !== value.contents.length ; ++idx) {
        if (typeof value.contents[idx] === 'object') {
          const v = value.contents[idx]; // v is {contents: ...}
          // We append the row context
          if (source.canShrinkWithoutContext(v)) {
            // As v will be passed towards the original shrinker during the possible shrinks on it (calling default)
            // We need to ensure it can be shrunk context-less (undefined being no context)
            shrs = shrs.join(fc.Stream.of(new fc.NextValue(v, undefined)));
          }
        }
      }
      // Fallback on the default shrinker
      shrs = shrs.join(defaultShrink());
      return shrs;
    },
  ),
})).record;

// Test

const includesBadValue = (c) => typeof c === "number" ? c === 50 : c.contents.some(includesBadValue)
fc.assert(fc.property(recordArb, record => !includesBadValue(record)), {numRuns: 10000, verbose: 2})
```

Live demo: https://runkit.com/dubzzz/60b4870b031969001a48a4f2

**Comment by msteen** [comment]
This is looking great! I have tested it locally and it works. Your example code does not type check for me (the helper), some `unknown` type gets injected, which requires quite a bit extra types to circumvent, so I just added a few `@ts-ignore` comments for now. I look forward to experimenting with this new version to see how I can improve my tests! I will be closing the issue as it should be addressed now. Thank you!
